Components/CommonCore/Source/gov/sandia/cognition/util/TwoWayIndex.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * File:                TwoWayIndex.java
+ * Authors:             Jeremy D. Wendt
+ * Company:             Sandia National Laboratories
+ * Project:             Cognitive Foundry
+ * 
+ * Copyright Apr 17, 2008, Sandia Corporation.
+ * Under the terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+ * license for use of this work by or on behalf of the U.S. Government. 
+ * Export of this program may require a license from the United States
+ * Government. See CopyrightHistory.txt for complete details.
+ * 
+ */
+
+package gov.sandia.cognition.util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Performs an indexing operation for a set of objects -- objects are assigned
+ * an index and the index can be used to map to the objects.
+ *
+ * @author jdwendt
+ *
+ * @param <ObjectType> The object type stored in the index
+ */
+public final class TwoWayIndex<ObjectType>
```

Comment:
I had to add a "clear" method to Indexer/DefaultIndexer and then removed TwoWayIndex.  You'll see it when I complete all other changes and update the pull request.